ocaml/xapi/stream_vdi.ml

```diff
@@ -100,6 +100,53 @@ let write_block ~__context filename buffer ofd len =
     then raise (Api_errors.Server_error (Api_errors.client_error, [ExnHelper.string_of_exn e]))
     else raise e
 
+(* Example = [{"length": 2097152, "flags": 0}, {"length": 1902848, "flags": 3}]*)
+type descriptor = {
+  length : int64;
+  status_flags : int32;
+}
+
+let cycle_descriptors descriptor_list offset =
+  let rec range a b =
+    if a > b then []
+    else a :: range (a + 1) b
+  in
+  (* Insert proper not_empty function here *)
+  let not_empty status_flag =
+    match (Int32.to_int status_flag) with
+      | 3 -> true
+      | _ -> false
+  in
+  let find_blocks descriptor offset increment =
+    match (not_empty descriptor.status_flags) with
+        | true -> begin
+            let start_chunk = offset / increment in
+            (* If the chunk ends at the boundary don't include the next chunk *)
+            let end_chunk = (Int64.to_int descriptor.length + offset - 1) / increment in
+            let chunks = range start_chunk end_chunk in
+            chunks, descriptor.length
+            end
+        | false -> [], descriptor.length
+  in
+  let check_exists a b =
+    if List.exists ((=)a) b = false then b@[a]
```

Comment:
Could this be written with `List.mem`? (`List.mem a b`). The name could be improved: what is being checked here?

Reply:
Discussed with Gabor/Stefano that only need to check it doesn't match the last value added to the accumulator rather than comparing against the whole list so will update.